pkg/cli/alpha/config-gen/types.go

```diff
@@ -0,0 +1,192 @@
+
+	// Status is the configuration status defined at runtime.
+	Status KubebuilderConfigGenStatus `json:"status,omitempty" yaml:"status,omitempty"`
+
+	// Project is for the field name used by the PROJECT file (config), instead of spec.
+	// Use Spec unless in the PROJECT file.
+	Project *KubebuilderConfigGenSpec `json:"config,omitempty" yaml:"config,omitempty"`
+}
+
+var (
+	// NamespaceComponent will create the namespace for the controller-manager
+	NamespaceComponent = "namespace"
+
+	// ControllerManagerComponent will create the controller-manager Deployment
+	ControllerManagerComponent = "controller-manager"
+
+	// WebhooksComponent will create the webhook configurations
+	WebhooksComponent = "webhooks"
+
+	// CRDsComponent will create the CRDs
+	CRDsComponent = "crds"
+
+	// RBACComponent will create RBAC rules
+	RBACComponent = "rbac"
+
+	// CertManagerComponent will create the Issuer and Certificate resources
+	// for CertManager to inject the certificates
+	CertManagerComponent = "cert-manager"
+
+	// PrometheusComponent will create the ServiceMonitor resource
+	PrometheusComponent = "prometheus"
+)
+
+// getDefaultComponents returns the set of components that are created by default
+func getDefaultComponents() map[string]bool {
+	return map[string]bool{
+		NamespaceComponent:         true,
+		ControllerManagerComponent: true,
+		RBACComponent:              true,
+		CRDsComponent:              true,
+	}
+}
+
+// Enabled returns true if the component is enabled
+func (a *KubebuilderConfigGenSpec) Enabled(component string) bool {
+	return a.Components[component]
+}
+
+// ComponentConfigEnabled returns true if component config is being used
+func (a *KubebuilderConfigGenSpec) ComponentConfigEnabled() bool {
+	return a.ComponentConfigFilepath != ""
+}
+
+// KubebuilderConfigGenSpec defines the desired configuration to be generated
+type KubebuilderConfigGenSpec struct {
+	// Directory is the kubebuilder directory containing the code.
+	// Defaults to the directoring containing the KubebuilderConfigGen configuration file.
+	Directory string `json:"projectDirectory" yaml:"projectDirectory"`
+
+	// Name is the name of project and used to generate the component and role names
+	// Defaults to metadata.name
+	Name string `json:"projectName" yaml:"projectName"`
+
+	// Namespace is the namespace to run the project in.
+	// Defaults to spec.name-system
+	Namespace string `json:"namespace" yaml:"namespace"`
+
+	// Image is the container image to run in the controller-manager Deployment
+	Image string `json:"image" yaml:"image"`
+
+	// Components is a map of components to enable or disable.
+	// Defaults to enabling:  ["namespace", "controller-manager", "rbac", "crds"]
+	Components map[string]bool `json:"components" yaml:"components"`
+
+	// DisableAuthProxy if set to true will disable the auth proxy
+	DisableAuthProxy bool `json:"disableAuthProxy,omitempty" yaml:"disableAuthProxy,omitempty"`
```

Comment:
Again, does it make sense to have it special-cased? I understand that this is a component that requires explicit disbaling instead of enabling, but probably a map makes more sense too. Would have to be checked in the `Enabled` method.

Reply:
This is a bit different from components, it's more about configuring components.  Either map or struct is fine.